File: EndGate/tests/EndGate.Core.Client.JS.Tests/Tests/UnitTests/Collision/CollisionManagerFacts.js
/// <reference path="../../Assets.ts" />

(function (window, collision, assets, bo) {

    QUnit.module("Collision Manager Facts");

    QUnit.test("Collision manager detects collisions.", function () {
        var cm = new collision.CollisionManager(),
            rect = new collision.Collidable(new bo.BoundingRectangle(new assets.Size2d(30, 50))),
            circle = new collision.Collidable(new bo.BoundingCircle(10)),
            rectCollisionTriggered = false,
            circleCollisionTriggered = false,
            managerCollisionTriggered = false;

        cm.Monitor(rect);
        cm.Monitor(circle);

        rect.OnCollision.Bind(function(collisionData) {
            QUnit.ok(rect.Bounds.Position.Equivalent(collisionData.At));
            QUnit.ok(circle === collisionData.With);
            rectCollisionTriggered = true;
        });

        circle.OnCollision.Bind(function(collisionData) {
            QUnit.ok(circle.Bounds.Position.Equivalent(collisionData.At));
            QUnit.deepEqual(rect, collisionData.With);
            circleCollisionTriggered = true;
        });
        
        cm.OnCollision.Bind(function (first, second) {
            QUnit.isSet(first);
            QUnit.isSet(second);
            QUnit.deepEqual(first, rect);
            QUnit.deepEqual(second, circle);
            managerCollisionTriggered = true;
        });       

        cm.Update(null);

        QUnit.ok(rectCollisionTriggered);
        QUnit.ok(circleCollisionTriggered);
        QUnit.ok(managerCollisionTriggered);

        rectCollisionTriggered = false;
        circleCollisionTriggered = false;
        managerCollisionTriggered = false;

        cm.Unmonitor(circle);
        cm.Update(null);

        QUnit.ok(!rectCollisionTriggered);
        QUnit.ok(!circleCollisionTriggered);
        QUnit.ok(!managerCollisionTriggered);

        cm.Monitor(circle);
        cm.Update(null);

        QUnit.ok(rectCollisionTriggered);
        QUnit.ok(circleCollisionTriggered);
        QUnit.ok(managerCollisionTriggered);

        rectCollisionTriggered = false;
        circleCollisionTriggered = false;
        managerCollisionTriggered = false;

        circle.Dispose();

        cm.Update(null);

        QUnit.ok(!rectCollisionTriggered);
        QUnit.ok(!circleCollisionTriggered);
        QUnit.ok(!managerCollisionTriggered);
    });

    QUnit.asyncTimeoutTest("Collision manager works within game.", testUtilities.defaultTestTimeout, function (end, assert, testName) {
        var rect = new collision.Collidable(new bo.BoundingRectangle(new assets.Size2d(30, 50))),
            circle = new collision.Collidable(new bo.BoundingCircle(10)),
            game = new CollisionManagerGame(),
            rectCollisionTriggered = 0,
            circleCollisionTriggered = 0,
            managerCollisionTriggered = 0;

        game.MonitorCollision(rect);
        game.MonitorCollision(circle);        

        rect.OnCollision.Bind(function(collisionData){
            assert.ok(rect.Bounds.Position.Equivalent(collisionData.At));
            assert.deepEqual(circle, collisionData.With);
            rectCollisionTriggered++;
        });

        circle.OnCollision.Bind(function(collisionData) {
            assert.ok(circle.Bounds.Position.Equivalent(collisionData.At));
            assert.deepEqual(rect, collisionData.With);
            circleCollisionTriggered++;
        });

        game.RegisterCollisionEvent(function(first, second) {
            assert.isSet(first);
            assert.isSet(second);
            assert.deepEqual(first, rect);
            assert.deepEqual(second, circle);
            managerCollisionTriggered++;
        });

        window.setTimeout(function () {
            var savedRectCollisions,
                savedCircleCollisions,
                savedmanagerCollisions;

            assert.ok(rectCollisionTriggered > 10);
            assert.ok(circleCollisionTriggered > 10);
            assert.ok(managerCollisionTriggered > 10);

            rect.Dispose();
            savedRectCollisions = rectCollisionTriggered;
            savedCircleCollisions = circleCollisionTriggered;
            savedmanagerCollisions = managerCollisionTriggered;

            window.setTimeout(function () {
                assert.deepEqual(savedRectCollisions, rectCollisionTriggered);
                assert.deepEqual(savedCircleCollisions, circleCollisionTriggered);
                assert.deepEqual(savedmanagerCollisions, managerCollisionTriggered);
                end();
            }, 1000);
        }, 1000);

        return function () {
        };
    });

})(window, EndGate.Core.Collision, EndGate.Core.Assets, EndGate.Core.BoundingObject);

File: EndGate/tests/EndGate.Core.Net.Tests/Collision/CollisionManagerFacts.cs
using System;
using System.Threading;
using EndGate.Core.Net.Assets;
using EndGate.Core.Net.BoundingObject;
using EndGate.Core.Net.Collision;
using Xunit;

namespace EndGate.Core.Net.Tests
{
    public class CollisionManagerFacts
    {
        [Fact]
        public void CollisionManagerDetectsCollisions()
        {
            var cm = new CollisionManager();
            var rect = new Collidable(new BoundingRectangle(new Size2d(30, 50)));
            var circle = new Collidable(new BoundingCircle(10));
            bool rectCollisionTriggered = false;
            bool circleCollisionTriggered = false;
            bool managerCollisionTriggered = false;

            cm.Monitor(rect);
            cm.Monitor(circle);

            rect.OnCollision += collisionData =>
            {
                Assert.True(rect.Bounds.Position.Equivalent(collisionData.At));
                Assert.Equal(circle, collisionData.With);
                rectCollisionTriggered = true;
            };

            circle.OnCollision += collisionData =>
            {
                Assert.True(circle.Bounds.Position.Equivalent(collisionData.At));
                Assert.Equal(rect, collisionData.With);
                circleCollisionTriggered = true;
            };

            cm.OnCollision += (first, second) =>
            {
                Assert.NotNull(first);
                Assert.NotNull(second);
                Assert.Equal(first, rect);
                Assert.Equal(second, circle);
                managerCollisionTriggered = true;
            };

            cm.Update(null);

            Assert.True(rectCollisionTriggered);
            Assert.True(circleCollisionTriggered);
            Assert.True(managerCollisionTriggered);

            rectCollisionTriggered = false;
            circleCollisionTriggered = false;
            managerCollisionTriggered = false;

            cm.Unmonitor(circle);
            cm.Update(null);

            Assert.False(rectCollisionTriggered);
            Assert.False(circleCollisionTriggered);
            Assert.False(managerCollisionTriggered);

            cm.Monitor(circle);
            cm.Update(null);

            Assert.True(rectCollisionTriggered);
            Assert.True(circleCollisionTriggered);
            Assert.True(managerCollisionTriggered);

            rectCollisionTriggered = false;
            circleCollisionTriggered = false;
            managerCollisionTriggered = false;

            circle.Dispose();

            cm.Update(null);

            Assert.False(rectCollisionTriggered);
            Assert.False(circleCollisionTriggered);
            Assert.False(managerCollisionTriggered);
        }

        [Fact]
        public void CollisionManagerWorksWithinGame()
        {
            var rect = new Collidable(new BoundingRectangle(new Size2d(30, 50)));
            var circle = new Collidable(new BoundingCircle(10));
            var game = new CollisionManagerGame();

            game.MonitorCollision(rect);
            game.MonitorCollision(circle);

            int rectCollisionTriggered = 0;
            int circleCollisionTriggered = 0;
            int managerCollisionTriggered = 0;

            rect.OnCollision += collisionData =>
            {
                Assert.True(rect.Bounds.Position.Equivalent(collisionData.At));
                Assert.Equal(circle, collisionData.With);
                rectCollisionTriggered++;
            };

            circle.OnCollision += collisionData =>
            {
                Assert.True(circle.Bounds.Position.Equivalent(collisionData.At));
                Assert.Equal(rect, collisionData.With);
                circleCollisionTriggered++;
            };

            game.RegisterCollisionEvent((first, second) =>
            {
                Assert.NotNull(first);
                Assert.NotNull(second);
                Assert.Equal(first, rect);
                Assert.Equal(second, circle);
                managerCollisionTriggered++;
            });

            Thread.Sleep(TimeSpan.FromSeconds(1));

            Assert.True(rectCollisionTriggered > 10);
            Assert.True(circleCollisionTriggered > 10);
            Assert.True(managerCollisionTriggered > 10);

            rect.Dispose();
            var savedRectCollisions = rectCollisionTriggered;
            var savedCircleCollisions = circleCollisionTriggered;
            var savedmanagerCollisions = managerCollisionTriggered;

            Thread.Sleep(TimeSpan.FromSeconds(1));

            Assert.Equal(savedRectCollisions, rectCollisionTriggered);
            Assert.Equal(savedCircleCollisions, circleCollisionTriggered);
            Assert.Equal(savedmanagerCollisions, managerCollisionTriggered);
        }

        private class CollisionManagerGame : Game
        {
            public void MonitorCollision(Collidable obj)
            {
                CollisionManager.Monitor(obj);
            }

            public void RegisterCollisionEvent(Action<Collidable, Collidable> e)
            {
                CollisionManager.OnCollision += e;
            }
        }
    }
}
